### PR TITLE
[REVIEW] add support for list_topic for a specific topic

### LINKED ISSUE #6351: [FEA] Add ability to specify specific Kafka topics in "list_topics()"
Add the ability to specifiy a specific Kafka topic in which a metadata request to "list_topics()" should retrieve.

### PR DESCRIPTION
**Pull request by jdye64**
This closes #6351

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

File: python/cudf_kafka/cudf_kafka/_lib/kafka.pyx [hunk 1]
```diff
@@ -47,9 +47,9 @@ cdef class KafkaDatasource(Datasource):
         return (<kafka_consumer *> self.c_datasource.get()). \
             get_committed_offset(topic, partition)
 
-    cpdef map[string, vector[int32_t]] list_topics(self):
+    cpdef map[string, vector[int32_t]] list_topics(self, string topic):
```
Comment (shwina) [inline python/cudf_kafka/cudf_kafka/_lib/kafka.pyx:50]: Should mark this `except +*` if it can raise C++ and/or Python exceptions.

File: python/cudf_kafka/cudf_kafka/_lib/kafka.pxd [hunk 2]
```diff
@@ -65,7 +65,7 @@ cdef class KafkaDatasource(Datasource):
 
     cpdef int64_t get_committed_offset(self, string topic, int32_t partition)
 
-    cpdef map[string, vector[int32_t]] list_topics(self)
+    cpdef map[string, vector[int32_t]] list_topics(self, string specific_topic)
```
Comment (shwina) [inline python/cudf_kafka/cudf_kafka/_lib/kafka.pxd:68]: Should mark this `except +*` if it can raise C++ and/or Python exceptions.

File: python/cudf_kafka/cudf_kafka/_lib/kafka.pxd [hunk 3]
```diff
@@ -34,7 +34,7 @@ cdef extern from "kafka_consumer.hpp" \
         int64_t get_committed_offset(string topic,
                                      int32_t partition) except +
 
-        map[string, vector[int32_t]] list_topics(string topic) except +
+        map[string, vector[int32_t]] list_topics(string topic) except +*
```
Comment (shwina) [inline python/cudf_kafka/cudf_kafka/_lib/kafka.pxd:37]: Since this is a binding to an external C++ function, it can only possibly raise a C++ exception. Thus, I think we should mark it `except +`.

File: python/cudf_kafka/cudf_kafka/_lib/kafka.pxd [hunk 4]
```diff
@@ -65,7 +65,7 @@ cdef class KafkaDatasource(Datasource):
 
     cpdef int64_t get_committed_offset(self, string topic, int32_t partition)
 
-    cpdef map[string, vector[int32_t]] list_topics(self, string specific_topic)
+    cpdef map[string, vector[int32_t]] list_topics(self, string topic) except +
```
Comment (shwina) [inline python/cudf_kafka/cudf_kafka/_lib/kafka.pxd:68]: Since this is a Python method that calls a C++ function, it can raise both Python and C++ exceptions. We should mark this one `except +*`.

Although, it's also OK to just mark it `except *`, since the C++ exception will already have been converted to a Python exception because we marked the `list_topics` C++ function `except +`.

I hope I'm making sense :)

File: python/cudf_kafka/cudf_kafka/_lib/kafka.pxd (hunk 3 above)
Comment (shwina) [reply python/cudf_kafka/cudf_kafka/_lib/kafka.pxd:37]: `except +` ensures that C++ exceptions get translated to Python exceptions.

File: cpp/libcudf_kafka/include/cudf_kafka/kafka_consumer.hpp [hunk 5]
```diff
@@ -149,9 +149,11 @@ class kafka_consumer : public cudf::io::datasource {
   /**
    * @brief Query the Kafka broker for the assigned topic Metadata
    *
+   * @param[in] Retrieves metadata for only a specific topic.
```
Comment (harrism) [inline cpp/libcudf_kafka/include/cudf_kafka/kafka_consumer.hpp:152]: @param requires a parameter name.  Also, maybe the name of this parameter should just be `topic_name`?

```suggestion
   * @param[in] specific_topic The name of the topic for which to retrieve metadata
```

File: cpp/libcudf_kafka/include/cudf_kafka/kafka_consumer.hpp [hunk 6]
```diff
@@ -149,9 +149,11 @@ class kafka_consumer : public cudf::io::datasource {
...
+   *
    * @return Map of Kafka topics with their corresponding topic partition values.
    */
-  std::map<std::string, std::vector<int32_t>> list_topics() const;
+  std::map<std::string, std::vector<int32_t>> list_topics(std::string specific_topic) const;
```
Comment (harrism) [inline cpp/libcudf_kafka/include/cudf_kafka/kafka_consumer.hpp:156]: If you are only getting a single topic, why does this need to return a map?

Comment (harrism) [reply cpp/libcudf_kafka/include/cudf_kafka/kafka_consumer.hpp:156]: If this only gets metadata for a single topic, why is it called "list topics"?

File: cpp/libcudf_kafka/src/kafka_consumer.cpp [hunk 7]
```diff
@@ -150,21 +150,29 @@ int64_t kafka_consumer::get_committed_offset(std::string const &topic, int parti
...
   int64_t offset = toppar_list[0]->offset();
-  return offset > 0 ? offset : 0;
+  return offset > 0 ? offset : -1001;
 }
 
-std::map<std::string, std::vector<int32_t>> kafka_consumer::list_topics() const
+std::map<std::string, std::vector<int32_t>> kafka_consumer::list_topics(
+  std::string specific_topic) const
 {
   auto const md = [&]() {
```
Comment (harrism) [inline cpp/libcudf_kafka/src/kafka_consumer.cpp:159]: With `auto` type and indecipherable name, it's unclear to the reader what this is. Let's give it a clearer name.

```suggestion
  auto const metadata = [&]() {
```

**Review by harrism — CHANGES_REQUESTED** [review]
Also missing gtest for this new feature. In fact, I can't seem to find any gtests for kafka in libcudf?

File: python/cudf_kafka/cudf_kafka/_lib/kafka.pyx [hunk 8]
```diff
@@ -47,9 +47,9 @@ cdef class KafkaDatasource(Datasource):
         return (<kafka_consumer *> self.c_datasource.get()). \
             get_committed_offset(topic, partition)
 
-    cpdef map[string, vector[int32_t]] list_topics(self):
+    cpdef map[string, vector[int32_t]] list_topics(self, string topic) except+:
```
Comment (shwina) [inline python/cudf_kafka/cudf_kafka/_lib/kafka.pyx:50]:
```suggestion
    cpdef map[string, vector[int32_t]] list_topics(self, string topic) except +*:
```

Needs to match the declaration in kafka.pxd

File: python/cudf_kafka/cudf_kafka/_lib/kafka.pxd [hunk 9]
```diff
@@ -34,7 +34,7 @@ cdef extern from "kafka_consumer.hpp" \
         int64_t get_committed_offset(string topic,
                                      int32_t partition) except +
 
-        map[string, vector[int32_t]] list_topics() except +
+        map[string, vector[int32_t]] list_topics(string topic) except +*
```
Comment (shwina) [inline python/cudf_kafka/cudf_kafka/_lib/kafka.pxd:37]:
```suggestion
        map[string, vector[int32_t]] list_topics(string topic) except +
```

Should just be `except +`, since this function can never raise Python exceptions

File: cpp/libcudf_kafka/include/cudf_kafka/kafka_consumer.hpp (hunk 6 above)
Comment (jdye64) [reply cpp/libcudf_kafka/include/cudf_kafka/kafka_consumer.hpp:156]: Since custreamz extends Python Streamz we are bound by previously defined function names to ensure our library works with theirs. This just happens to be one of those functions names that maps to what they have and we can't change.

Comment (jdye64) [reply cpp/libcudf_kafka/include/cudf_kafka/kafka_consumer.hpp:156]: I guess on the C++ side we could change the name and leave the python name as is but that almost seemed more confusing. I can do that if desired however.

Comment (harrism) [reply cpp/libcudf_kafka/include/cudf_kafka/kafka_consumer.hpp:156]: In libcudf we always try to do the right thing rather than doing the wrong thing because somebody else does it.  On the C++ side we have the freedom to do the right thing.

In the long term I think it's better to build a clear and usable API. If we don't, we end up spending way more refactoring later.

Comment (chinmaychandak) [reply cpp/libcudf_kafka/include/cudf_kafka/kafka_consumer.hpp:156]: > If this only gets metadata for a single topic, why is it called "list topics"?

If no topic name is specified, it gets the metadata for all topics. But we need to provide functionality to retrieve metadata for specific topics as well, since the metadata for all topics might not be needed at all times and can be a pretty large object at times.

The name `list_topics` is the name of the standard Confluent Kafka Python library function, which `custreamz.kafka` or `cudf_kafka` serves to be a drop-in replacement for in order to use GPUs instead of CPUs.

Comment (harrism) [reply cpp/libcudf_kafka/include/cudf_kafka/kafka_consumer.hpp:156]: This is not what the function documentation says. Why not make the function signature like this:

```
std::map<std::string, std::vector<int32_t>> list_topics(std::vector<std::string> const& topics) const;
```

Then the user can query a subset of topics (not just one or all). Document that if `topics` is empty, metadata for all topics is returned.  This API will be no more difficult to use, since the user can just call `list_topics({topic_name})` to convert a single name to a vector.

Comment (jdye64) [reply cpp/libcudf_kafka/include/cudf_kafka/kafka_consumer.hpp:156]: I will name it whatever the group wants to name it. That part really doesn't matter to me. @harrism if this was coming from scratch I love your idea about a `std::vector` however the way the underlying C API is written right now only a single topic_name at a time is accepted https://docs.confluent.io/2.0.0/clients/librdkafka/classRdKafka_1_1Handle.html#ac0cc63c574ce027a184640645151a75c Now that isn't a big deal as I could of course iterate over the std::vector of topics and call that function multiple times but that isn't how folks using the API currently do it.

In a nut shell doesn't matter to me just wanted to lay out all the facts. Let me know which you prefer and I will take that approach.

Comment (harrism) [reply cpp/libcudf_kafka/include/cudf_kafka/kafka_consumer.hpp:156]: OK, I think leaving the API is OK, but please improve the documentation to be clear and accurate. Thanks!

File: python/cudf_kafka/cudf_kafka/_lib/kafka.pxd [hunk 10]
```diff
@@ -65,7 +65,7 @@ cdef class KafkaDatasource(Datasource):
 
     cpdef int64_t get_committed_offset(self, string topic, int32_t partition)
 
-    cpdef map[string, vector[int32_t]] list_topics(self)
+    cpdef map[string, vector[int32_t]] list_topics(self, string tp) except +*
```
Comment (shwina) [inline python/cudf_kafka/cudf_kafka/_lib/kafka.pxd:68]: Not sure why Cython is crashing, but tested some similar locally and this seems to work for me:

```suggestion
    cpdef map[string, vector[int32_t]] list_topics(self, string tp) except *
```

Also needs a similar change in `kafka.pyx`

File: python/cudf_kafka/cudf_kafka/_lib/kafka.pyx [hunk 11]
```diff
@@ -47,9 +47,10 @@ cdef class KafkaDatasource(Datasource):
         return (<kafka_consumer *> self.c_datasource.get()). \
             get_committed_offset(topic, partition)
 
-    cpdef map[string, vector[int32_t]] list_topics(self):
+    cpdef map[string, vector[int32_t]] list_topics(self,
+                                                   string topic) except+*:
```
Comment (shwina) [inline python/cudf_kafka/cudf_kafka/_lib/kafka.pyx:51]: See my comment above

```suggestion
                                                   string topic) except *:
```

**Comment by jdye64** [comment]
rerun tests

**Review by shwina — APPROVED** [review]
Python LGTM!

File: cpp/libcudf_kafka/include/cudf_kafka/kafka_consumer.hpp [hunk 12]
```diff
@@ -147,11 +147,13 @@ class kafka_consumer : public cudf::io::datasource {
...
+   * specified then the partitions for all Topics in the broker will be retrieved.
    *
-   * @param[in] specific_topic The name of the topic for which to retrieve metadata
+   * @param[in] specific_topic The name of the topic for which to retrieve partitions. If empty then
+   * the partitions for all topics will be retrieved.
    *
-   * @return Map of Kafka topics with their corresponding topic partition values.
+   * @return Map of Kafka topic names with their corresponding list of topic partition values.
    */
   std::map<std::string, std::vector<int32_t>> list_topics(std::string specific_topic) const;
```
Comment (harrism) [inline cpp/libcudf_kafka/include/cudf_kafka/kafka_consumer.hpp:158]: Perhaps `specific_topic` should default to the empty string to make this easy to call when you want all topics?

```suggestion
  std::map<std::string, std::vector<int32_t>> list_topics(std::string specific_topic = std::string{}) const;
```

**Review by harrism — APPROVED** [review]
Approve, but suggest a default argument.

**Comment by chinmaychandak** [comment]
@jdye64, do we need any more changes here?

File: cpp/libcudf_kafka/src/kafka_consumer.cpp [hunk 13]
```diff
@@ -150,21 +150,29 @@ int64_t kafka_consumer::get_committed_offset(std::string const &topic, int parti
                "Failed retrieve Kafka committed offsets");
 
   int64_t offset = toppar_list[0]->offset();
-  return offset > 0 ? offset : 0;
+  return offset > 0 ? offset : -1001;
```
Comment (vuule) [inline cpp/libcudf_kafka/src/kafka_consumer.cpp:153]: Is this change intentional? Seems unrelated to the API change the PR is about.

File: cpp/libcudf_kafka/src/kafka_consumer.cpp [hunk 14]
```diff
@@ -150,21 +150,29 @@ int64_t kafka_consumer::get_committed_offset(std::string const &topic, int parti
...
+std::map<std::string, std::vector<int32_t>> kafka_consumer::list_topics(
+  std::string specific_topic) const
 {
-  auto const md = [&]() {
+  auto const metadata = [&]() {
+    std::string errstr;
+    RdKafka::Topic const *const spec_topic =
+      specific_topic.empty()
+        ? nullptr
+        : RdKafka::Topic::create(consumer.get(), specific_topic, nullptr, errstr);
```
Comment (vuule) [inline cpp/libcudf_kafka/src/kafka_consumer.cpp:164]: Do we need to check `errstr` after this call?

File: cpp/libcudf_kafka/src/kafka_consumer.cpp (hunk 13 above)
Comment (chinmaychandak) [reply cpp/libcudf_kafka/src/kafka_consumer.cpp:153]: Yes, we wanted to change the default value for this API, and thought we could just squeeze this in.

File: cpp/libcudf_kafka/src/kafka_consumer.cpp [hunk 15]
```diff
@@ -150,21 +150,29 @@ int64_t kafka_consumer::get_committed_offset(std::string const &topic, int parti
...
+
     RdKafka::Metadata *md;
-    CUDF_EXPECTS(RdKafka::ERR_NO_ERROR == consumer->metadata(true, nullptr, &md, default_timeout),
-                 "Failed to list_topics in Kafka broker");
+    CUDF_EXPECTS(
+      RdKafka::ERR_NO_ERROR ==
+        consumer->metadata(specific_topic.empty() ? true : false, spec_topic, &md, default_timeout),
```
Comment (vuule) [inline cpp/libcudf_kafka/src/kafka_consumer.cpp:169]: I think with this change you don't need to check the error string after the `create` call. Unless we want the `list_topics` call to fail if something goes wrong.
I recommend this change either way. IMO it's cleaner since we now don't repeat the `specific_topic.empty()` check.
```suggestion
        consumer->metadata(spec_topic == nullptr ? true : false, spec_topic, &md, default_timeout),
```

File: cpp/libcudf_kafka/src/kafka_consumer.cpp [hunk 16]
```diff
@@ -150,21 +150,29 @@ int64_t kafka_consumer::get_committed_offset(std::string const &topic, int parti
...
 }
 
-std::map<std::string, std::vector<int32_t>> kafka_consumer::list_topics() const
+std::map<std::string, std::vector<int32_t>> kafka_consumer::list_topics(
+  std::string specific_topic) const
 {
-  auto const md = [&]() {
+  auto const metadata = [&]() {
+    std::string errstr;
+    RdKafka::Topic const *const spec_topic =
```
Comment (vuule) [inline cpp/libcudf_kafka/src/kafka_consumer.cpp:161]: Does `spec_topic` need to be manually deleted? I can't find this info in their docs.

**Review by vuule — CHANGES_REQUESTED** [review]
Found a few potential issues.

File: cpp/libcudf_kafka/src/kafka_consumer.cpp (hunk 15 above)
Comment (jdye64) [reply cpp/libcudf_kafka/src/kafka_consumer.cpp:169]: I like this and think its the best way and prevents the extra check on the `errstr` after the call the create. I'll make the change

File: cpp/libcudf_kafka/src/kafka_consumer.cpp (hunk 16 above)
Comment (jdye64) [reply cpp/libcudf_kafka/src/kafka_consumer.cpp:161]: Yes it does. I'll do that.

File: cpp/libcudf_kafka/src/kafka_consumer.cpp [hunk 17]
```diff
@@ -150,21 +150,27 @@ int64_t kafka_consumer::get_committed_offset(std::string const &topic, int parti
...
+  auto const metadata = [&]() {
+    std::string errstr;
+    auto spec_topic = std::unique_ptr<RdKafka::Topic>(
+      RdKafka::Topic::create(consumer.get(), specific_topic, nullptr, errstr));
+
     RdKafka::Metadata *md;
-    CUDF_EXPECTS(RdKafka::ERR_NO_ERROR == consumer->metadata(true, nullptr, &md, default_timeout),
+    CUDF_EXPECTS(RdKafka::ERR_NO_ERROR ==
+                   consumer->metadata(
+                     spec_topic == nullptr ? true : false, spec_topic.get(), &md, default_timeout),
```
Comment (vuule) [inline cpp/libcudf_kafka/src/kafka_consumer.cpp:167]: Just realized this is completely redundant. Could be:
```suggestion
                     spec_topic == nullptr, spec_topic.get(), &md, default_timeout),
```

**Review by vuule — APPROVED** [review]

**Comment by harrism** [comment]
```
ImportError: /opt/conda/envs/rapids/conda-bld/custreamz_1601946400766/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/lib/python3.7/site-packages/cudf_kafka/_lib/kafka.cpython-37m-x86_64-linux-gnu.so: undefined symbol: _ZNK4cudf2io8external5kafka14kafka_consumer11list_topicsENSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE
```

Looks like some missing linker  settings for the Python bindings?

**Comment by jdye64** [comment]
> ```
> ImportError: /opt/conda/envs/rapids/conda-bld/custreamz_1601946400766/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/lib/python3.7/site-packages/cudf_kafka/_lib/kafka.cpython-37m-x86_64-linux-gnu.so: undefined symbol: _ZNK4cudf2io8external5kafka14kafka_consumer11list_topicsENSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE
> ```
> 
> Looks like some missing linker settings for the Python bindings?

Yeah ... strange. To quote the phrase from the ancient developers "it works on my machine". I'm sure it is something with the build order in the CI scripts. I'll take a look now.

**Comment by jdye64** [comment]
rerun tests

**Comment by ajschmidt8** [comment]
rerun tests

**Comment by chinmaychandak** [comment]
Any idea why this the CI is still failing?

**Comment by harrism** [comment]
Yeah as I mentioned above, there is a missing library, presumably in the setup.py configuration for the Cython bindings.

**Comment by jdye64** [comment]
> Yeah as I mentioned above, there is a missing library, presumably in the setup.py configuration for the Cython bindings.

After speaking with ops and Keith it’s actually a problem with anaconda not pulling the latest ‘libcudf_kafka’ build which leads to the missing symbols error since the older builds of ‘libcudf_kafka’ don’t have those functions

**Comment by chinmaychandak** [comment]
> it’s actually a problem with anaconda not pulling the latest ‘libcudf_kafka’ build which leads to the missing symbols error since the older builds of ‘libcudf_kafka’ don’t have those functions

Is there a PR open to fix this issue somehow? Seems weird that it isn't picking up the latest build of libcudf_kafka (but picks up latest builds of other RAPIDS packages), did you guys figure out why it might be happening?

**Comment by kkraus14** [comment]
rerun tests
